File: 1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero/1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero.java
class Solution {
    public int minReorder(int n, int[][] connections) {
        HashSet<String> set = new HashSet<>();

        List<List<Integer>> adj = new ArrayList<>();
        for(int i=0; i<n; i++) adj.add(new ArrayList<>());

        for(int[] edge : connections){
            set.add(edge[0] + " to " + edge[1]);

            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }

        int changes = 0;
        Queue<Integer> q = new LinkedList<>();
        boolean[] vis = new boolean[n];

        q.add(0);
        vis[0]=true;

        while(!q.isEmpty()){
            int curr = q.poll();

            for(int neigh : adj.get(curr)){
                if(!vis[neigh]){
                    if(set.contains(curr + " to " + neigh)){
                        changes++;
                    }
                    q.add(neigh);
                    vis[neigh] = true;
                }
            }
        }


        return changes;
    }
}

File: 1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero/1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero.py
class Solution(object):

    def dfs(self, i, routes, vis, adj):
        vis[i] = True

        for next_i in adj[i]:
            if not vis[next_i]:
                if (i, next_i) in routes:
                    self.res += 1
                self.dfs(next_i, routes, vis, adj)
                

    def minReorder(self, n, connections):
        """
        :type n: int
        :type connections: List[List[int]]
        :rtype: int
        """
        routes = set()
        adj = [[] for _ in range(n)]
        self.res = 0

        for edge in connections:
            routes.add((edge[0], edge[1]))
            adj[edge[0]].append(edge[1])
            adj[edge[1]].append(edge[0])

        visited = [False] * n
        
        self.dfs(0, routes, visited, adj)

        return self.res
